feat : 게임시작, 금액배팅 구현

gutenLEE opened the pull request:
1. 게임시작
- GameApp.java
- 배팅 금액 입력
- 게이머와 딜러 각자 받은 카드 공개

2. 게이머와 딜러가 받은 카드 공개 기능 인터페이스 생성 : Openable.java

Commit:
feat : 게임시작, 금액배팅 구현

1. 게임시작
- GameApp.java
- 배팅 금액 입력
- 게이머와 딜러 각자 받은 카드 공개

2. 게이머와 딜러가 받은 카드 공개 기능 인터페이스 생성 : Openable.java

f-lab-maverick commented on src/batting/Money.java:
```diff
@@ -5,6 +5,10 @@
  * @since 2023/06/18
  */
 public class Money {
+    private Integer battingMoney;
```
Wrapper type을 쓰신 이유가 있으실까요? 초기화가 되지 않았을 때 초기값이 null이라면 문제가 생기지 않을까요?

f-lab-maverick commented on src/card/Card.java:
```diff
@@ -27,4 +29,9 @@ private CardType getCardType(int number) {
+
+    @Override
+    public String toString() {
+        return String.format("[ %s, %s, %s ]", cardType, (cardType==NUMBER ? number : ""), suit);
```
카드 타입에 따라 모양을 내리는지 숫자를 내리는지를 결정하는 로직이 있는데요, 이 부분은 비즈니스 로직이라고 보면 좋을 것으로 보입니다.

`toString`에서만 사용하는 로직이 아니므로 이는 도메인에 공통으로 사용할 수 있는 메서드를 만드는것이 적절하겠네요.

단순한 출력 메서드에 비즈니스로직을 넣는 행위는 지양해야 합니다. 왜나하면 이런 비즈니스 로직은 반복해서 사용하게 될 것인데, 이를 메서드가 아닌 사용할 때 마다 중복으로 코드를 작성하게 되면 추후 요구사항이 변경되었을 때 모두 다 찾아가서 고쳐주어야 하는 문제가 생기고, 하나라도 놓칠 경우 버그로 이어질 수 있기 때문입니다.

f-lab-maverick commented on src/card/CardHolder.java:
```diff
@@ -20,4 +20,7 @@ public void addCard(Card card) {
 
+    public List<Card> getCardList() {
+        return cardList;
+    }
```
내부적으로 가지고있는 cardList를 그대로 리턴한다면 이 객체 외부에서 cardList를 조작할 수 있는 위험이 있지 않을까요?

ZhenxiKim replied:
외부 조작 위험으로 벗어나기 위해 clone() 메소드를 써서 객체의 독립성을 보장할 수 있을것 같습니다.

f-lab-maverick commented on src/user/Openable.java:
```diff
@@ -0,0 +1,6 @@
+package user;
+
+public interface Openable {
```
👍 추상화 시도 좋습니다

f-lab-maverick commented on src/user/Dealer.java:
```diff
@@ -1,15 +1,22 @@
 
+    @Override
+    public void openCard() {
+        Card card = cardHolder.getCardList().get(0);
```
`IndexOutOfBoundException`이 발생할 여지가 있습니다. 직접적으로 인덱스를 지정하여 List에서 Element를 꺼내는 행위는 지양해야 해요.

이런 경우, Java Stream에서 제공해주는 `findFirst`, `orElseThrow` 기능을 사용하여 예외를 명시적으로 처리할 수 있겠네요.

f-lab-maverick commented on src/user/Dealer.java:
```diff
@@ -1,15 +1,22 @@
+    @Override
+    public void openCard() {
+        Card card = cardHolder.getCardList().get(0);
+        System.out.println("Dealer Card [" + card + "]");
```
출력은 도메인의 역할이 아닙니다. 출력을 하는 부분은 모두 View 레이어에서 할 수 있도록 지정해주세요.

출력문을 View 영역이 아닌 도메인, 서비스레이어에서 진행한다면 추후 View와 소스코드의 결합도가 강하게 발생하고 이 출력이 어떻게 발생하는지 추적하기 어려워집니다.

f-lab-maverick commented on src/user/Gamer.java:
```diff
@@ -6,9 +6,18 @@
+
+    @Override
+    public void openCard() {
+        System.out.println("Gamer Card " + cardHolder.getCardList());
```
이 부분도 동일하게 View 레이어로 출력문을 변경해주세요.

f-lab-maverick replied:
또한 Delaler의 경우 한장의 카드만 open하는데 Gamer는 모든 카드를 오픈하네요. 동일한 인터페이스를 구현하는데 구현 내용이 차이가 나는것으로 보입니다.

필요하다면 `OpenFirst`, `OpenAll` 등으로 구분하여 구현하는것이 명시적이겠습니다.

f-lab-maverick commented on src/GameApp.java:
```diff
@@ -0,0 +1,37 @@
+        gamer.receiveCard(cardDeck.giveCard());
+        gamer.receiveCard(cardDeck.giveCard());
+        dealer.receiveCard(cardDeck.giveCard());
+        dealer.receiveCard(cardDeck.giveCard());
```
이후에 이런 중복코드는 메서드 or 반복문으로 변경해주세요~

f-lab-maverick commented on test/card/CardDeckTest.java:
```diff
@@ -0,0 +1,18 @@
+    void giveCardShouldNotNull(){
+        CardDeck cardDeck = CardDeck.create();
+        Card card = cardDeck.giveCard();
+        assertNotNull(card);
```
단언문 선언 좋습니다~ 테스트를 여러개 작성해보시고, `@DisplayName`으로 테스트하고자 하는 내용을 한글로도 자세히 작성해보세요

f-lab-maverick reviewed (changes requested):
추상화 시도 좋습니다. 다만, 여러가지 버그 발생 가능성이 있는 코드가 약간 보입니다. 이 부분을 수정해보시겠어요?

Commit:
feat : 게임시작, 금액배팅 구현 수정

1. View 레이어 생성
- CardViewer.java

2. CardHolder getter 메서드들의 리턴 참조 변수는 변경할 수 없도록함

3. Openable 인터페이스 제거
- 의도는 게이머와 딜러 모두 카드를 오픈하지만 오픈 카드 개수의 차이라 생각해 인터페이스를 사용했음. 룰을 잘못 이해하였음.

- 딜러만 갖는 행위인 카드 오픈은 딜러만 갖도록 변경

4. Money Wrapper 필드 초기화

f-lab-maverick commented on src/user/Dealer.java:
```diff
@@ -3,20 +3,21 @@
-        System.out.println("Dealer Card [" + card + "]");
+    public Card openFirstCard() {
+        return Optional.ofNullable(cardHolder.getFirstCard())
+                .orElseThrow(NoSuchElementException::new);
```
`NoSuchElementException`은 이렇게 명시적으로 throw하지 않아도 enum에서 `get`을 사용시 발생하는 Exception이라 이렇게 선언하지 않아도 괜찮습니다.

다만, 예외는 명시적으로 던지는게 좋습니다. 예외 메세지, Custom Exception등으로 어떤 예외가 발생했는지 명확히 표현해보세요

f-lab-maverick commented on src/batting/Money.java:
```diff
@@ -5,8 +5,8 @@
 public class Money {
-    private Integer battingMoney;
-    private Integer remains;
+    private Integer battingMoney = 0;
+    private Integer remains = 0;
```
제가 드린 피드백중 Wrapper Type을 왜 사용했는지가 있었는데요, 특별한 이유가 없다면 primitive type을 사용하는게 좋습니다. 아래 링크를 참고해보세요
https://www.baeldung.com/java-primitives-vs-objects

f-lab-maverick commented on src/card/CardHolder.java:
```diff
@@ -20,7 +19,20 @@ public void addCard(Card card) {
 
-    public List<Card> getCardList() {
-        return cardList;
+    public Card getFirstCard() {
+        if (hasCards()) {
+            return new Card(cardList.get(0));
+        }
+        return null;
```
null을 리턴하는 방식은 NPE를 발생시킬 위험이 상당히 큽니다. 카드를 리턴해야 하는데 카드를 꺼낼 수 없다면 이건 예외사항이므로 명시적으로 Exception을 발생시켜 예외처리를 하는것이 더더욱 깔끔한 코드 로직을 만들 수 있습니다. null을 리턴해야 하는 경우는 Optional을 리턴값으로 주어 이 리턴값이 비어있을 수 있음을 보여주는게 좋습니다. (이 코드는 예외처리가 더 적절해 보이네요)

f-lab-maverick commented on src/card/CardHolder.java:
```diff
@@ -20,7 +19,20 @@ public void addCard(Card card) {
+        return null;
     }
+
+    public List<Card> getAllCard() {
+        return Collections.unmodifiableList(cardList);
+    }
```
👍

f-lab-maverick commented on src/card/CardViewer.java:
```diff
@@ -0,0 +1,31 @@
+        int cardNumber = card.getNumber();
+        Suit suit = card.getSuit();
+
+        return String.format("[ %s, %s, %s ]", cardType, (cardType==NUMBER ? cardNumber : ""), suit);
```
`cardType==NUMBER ? cardNumber : ""`은 비즈니스로직에 해당합니다. 도메인 내부로 메서드를 옮길 수 없을지 고민해보시겠어요?

f-lab-maverick commented on test/card/CardHolderTest.java:
```diff
@@ -0,0 +1,31 @@
+
+class CardHolderTest {
+
+    @Test
+    @DisplayName("CardHolder 가 리턴한 카드 리스트는 수정할 수 없다")
+    void getAllCardCannotModify() throws Exception {
+
+        // given
+        CardHolder cardHolder = CardHolder.create();
+        cardHolder.addCard(new Card(1, Suit.HEART));
+        cardHolder.addCard(new Card(2, Suit.SPADE));
+
+        // when
+        List<Card> allCard = cardHolder.getAllCard();
+
+        // then
+        assertThrows(UnsupportedOperationException.class, () -> {
+            allCard.add(1, new Card(3, Suit.CLOVER));
+        });
+    }
```
명확한 테스트 좋습니다!

f-lab-maverick reviewed (changes requested):
고생하셨습니다. 피드백 확인 부탁드려요~
그리고 제 피드백을 확인했다면 질문형 피드백일 경우 답변을 적어주시고, 다른 경우에는 확인했다는 표시로 이모티콘 하나 남겨주시면 좋을듯 합니다~

gutenLEE commented:
> 



> 고생하셨습니다. 피드백 확인 부탁드려요~ 그리고 제 피드백을 확인했다면 질문형 피드백일 경우 답변을 적어주시고, 다른 경우에는 확인했다는 표시로 이모티콘 하나 남겨주시면 좋을듯 합니다~

앗 반영하겠습니다 감사합니다!

Commit:
feat : 게임시작, 금액배팅 구현 수정 2

1. CardHolderEmptyException 추가
- 카드 홀더에 카드가 없는데 카드 요청을 받을 경우 게임을 진행할 수 없는 상태이므로 예외 던지기

gutenLEE commented on src/card/Card.java:
```diff
@@ -33,6 +33,10 @@ private CardType getCardType(int number) {
         return type;
     }
 
+    public boolean isNumber(){
```
카드가 넘버인지 여부를 판별해 주는 거까지만 비즈니스 로직이고
넘버 여부에 따라 출력할 문자를 선택하는 것은 view 레이어의 책임이라 생각하여 이렇게 해보았습니다!

f-lab-maverick replied:
네 비즈니스 로직을 고민한 부분 좋습니다~
다만 문자의 출력이라는 부분이
1. 숫자카드인가?
2. 맞다면 숫자 출력
3. 아니면 문자 출력

이 부분이 반복될 수 있습니다. 추후에 이 부분이 중복으로 발생한다면 이 부분도 메서드로 변경하는걸 검토해보는게 좋을 것으로 보이네요

gutenLEE commented on src/GameApp.java:
```diff
@@ -31,7 +29,14 @@ public static void main(String[] args) {
+        Card openFirstCard = null;
+        try {
+            openFirstCard = dealer.openFirstCard();
+        } catch (CardHolderEmptyException e){
```
카드 홀더에 카드가 없을 경우는 giveCards() 에서부터 문제가 발생한 것이고
게임 프로세스가 더 이상 진행될 수 없는 예외 상황이기 때문에
runtimeException 상속한 CardHolderEmptyException 의 try-catch 를 GameApp 에서 선언하였습니다.

gutenLEE commented on src/card/CardViewer.java:
```diff
@@ -20,12 +19,15 @@ public static void print(List<Card> cardList) {
     }
 
-    private static String getCardInfo(Card card) {
+    public static String getCardInfo(Card card) {
```
테스트를 위해 private 을 public 으로 변경했습니다.
view 레이어의 getter 메서드라 크게 부수 효과는 없을거라 판단했습니다. 

찾아보니 private 메서드는 테스트 하지 않으며, 필요할 경우 reflection API 를 사용해 테스트할 수 있다... 로 결론을 얻게 되었습니다. 

이럴 경우 메서드의 리턴 결과를 테스트하기 보단, 메서드 구현부에 선언된 코드를 테스트하면 적절한지 궁금합니다.

yyy9942 replied:
private method를 테스트를 위해 public으로 변경하는건 좋은 방식은 아닙니다. 
테스트 대상을 지정하는 부분 때문에 private 메서드에 대한 테스트를 고민하시는거 같아요.  메서드를 테스트의 단위로 생각하게 된다면 private 메서드도 테스트 대상이 됩니다.

관련해서 명확하게 정답은 없지만, [Stackoverflow](https://stackoverflow.com/questions/105007/should-i-test-private-methods-or-only-public-ones)에서 여러 토론을 진행하였으니 한번 확인해보시겠어요?. 한번 읽어보시고 테스트의 범위와 대상을 어떻게 지정하는것이 좋을지 고민해본다면 좋은 경험이 될 것으로 보입니다 :)

f-lab-maverick replied:
private method를 테스트를 위해 public으로 변경하는건 좋은 방식은 아닙니다. 
테스트 대상을 지정하는 부분 때문에 private 메서드에 대한 테스트를 고민하시는거 같아요.  메서드를 테스트의 단위로 생각하게 된다면 private 메서드도 테스트 대상이 됩니다.

관련해서 명확하게 정답은 없지만, [Stackoverflow](https://stackoverflow.com/questions/105007/should-i-test-private-methods-or-only-public-ones)에서 여러 토론을 진행하였으니 한번 확인해보시겠어요?. 한번 읽어보시고 테스트의 범위와 대상을 어떻게 지정하는것이 좋을지 고민해본다면 좋은 경험이 될 것으로 보입니다 :)

f-lab-maverick commented on src/card/CardHolderEmptyException.java:
```diff
@@ -0,0 +1,8 @@
+ * @author gutenlee
+ * @since 2023/06/26
+ */
+public class CardHolderEmptyException extends RuntimeException {
```
👍

f-lab-maverick commented on test/user/DealerTest.java:
```diff
@@ -0,0 +1,43 @@
+class DealerTest {
+
+    @Test
+    @DisplayName("딜러가 보유한 카드가 있을 경우 openFirstCard 는 not null 이다. ")
```
있을경우는 card를 리턴하고, 없을 경우 Exception을 throw한다면 이 메서드는 무조건 not null이라고 볼 수도 있을것 같네요~

f-lab-maverick reviewed (approved):
고생하셨습니다~ 머지하셔도 좋습니다
